### PR TITLE
Add complete k8s metadata through composable provider

### PR DESCRIPTION
**Pull request by ChrsMark**
## What does this PR do?
Add all k8s metadata via the provider.

## Why is it important?
To support full metadata access like in Beats.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## TODOs

- [x] Verify that short-living pods are handled properly.

## How to test this PR locally
1. Enable a static autodiscovery rule in Agent's config:
```
providers.kubernetes:
  scope: node
  kube_config: /Users/chrismark/.kube/config
  node: "kind-control-plane"
  cleanup_timeout: 360s
  resources:
    pod:
      enabled: true

inputs:
- name: redis
  type: redis/metrics
  use_output: default
  meta:
    package:
      name: redis
      version: 0.3.6
  data_stream:
    namespace: default
  streams:
    - data_stream:
        dataset: redis.info
        type: metrics
      metricsets:
        - info
      hosts:
        - '${kubernetes.pod.ip}:${kubernetes.container.port}'
      idle_timeout: 20s
      maxconn: 10
      network: tcp
      period: 10s
      condition: ${kubernetes.labels.app} == 'redis' AND ${kubernetes.container.port_name} == 'web'
```
2. Deploy a Redis pod that triggers the condition:
```
apiVersion: v1
kind: Pod
metadata:
  name: redis
  labels:
    role: main
    app: redis
  annotations:
    # this had no meaning
    co.elastic.metrics.redis.XYZ/module: redis
spec:
  hostNetwork: true
  dnsPolicy: ClusterFirstWithHostNet
  containers:
    - name: redis
      image: redis:latest
      ports:
        - name: web
          containerPort: 6379
          protocol: TCP
```
3. Run agent's inspect command to verify if an input is created `./elastic-agent -c ./elastic-agent.yml inspect output -o default`
4. Verify that an input is created and metadata are added properly via the processors, for example:
```
metricbeat:
  modules:
  - hosts:
    - 172.18.0.3:6379
    idle_timeout: 20s
    index: metrics-redis.info-default
    maxconn: 10
    meta:
      package:
        name: redis
        version: 0.3.6
    metricsets:
    - info
    module: redis
    name: redis
    network: tcp
    period: 10s
    processors:
    - add_fields:
        fields:
          id: cda93b8385201eaf725532e908e4a4c55895273d2a2cb57700b3cf48732b2de3
          image:
            name: redis:latest
          runtime: containerd
        target: container
    - add_fields:
        fields:
          container:
            name: redis
          labels:
            app: redis
            role: main
          namespace: default
          node:
            name: kind-control-plane
          pod:
            ip: 172.18.0.3
            name: redis
            uid: 365e8d21-a81d-44e2-acbf-95deb8a57730
        target: kubernetes
    - add_fields:
        fields:
          cluster:
            name: kind-kind
            url: https://127.0.0.1:50740
        target: orchestrator
    - add_fields:
        fields:
          dataset: redis.info
          namespace: default
          type: metrics
        target: data_stream
    - add_fields:
        fields:
          dataset: redis.info
        target: event
    - add_fields:
        fields:
          id: 32806c0a-4f04-499c-9427-e6e24e5f6035
          snapshot: false
          version: 8.0.0
        target: elastic_agent
    - add_fields:
        fields:
          id: 32806c0a-4f04-499c-9427-e6e24e5f6035
        target: agent
```

### Using annotations (Note: annotations are not exposed to meta fields by default but are exposed in variable resolution mechanism)
Specify a condition based on an annotation like `condition: ${kubernetes.annotations.level} == 'production' AND ${kubernetes.container.port_name} == 'web'` and verify that input is created again. (note: tune Redis Pod manifest accordingly)

### Short-living pods
Define a log input with condition based on container's name:
```
      - name: container-log
        type: logfile
        use_output: default
        meta:
          package:
            name: log
            version: 0.4.6
        data_stream:
          namespace: default
        streams:
          - data_stream:
              dataset: generic
            symlinks: true
            condition: ${kubernetes.container.name} == 'hello'
            paths:
              - /var/log/containers/*${kubernetes.container.id}.log
```

Deploy elastic-agent and verify that it collects logs for the following short-living pod:
```
apiVersion: batch/v1
kind: CronJob
metadata:
  name: mytarge
  labels:
    app: mytarget
spec:
  schedule: "*/1 * * * *"
  failedJobsHistoryLimit: 10
  successfulJobsHistoryLimit: 20
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: hello
            image: busybox
            args:
            - /bin/sh
            - -c
            - date; echo Hello from the Kubernetes cluster
          restartPolicy: OnFailure
```
Screenshot 2021-09-15 at 3 56 18 PM

### Short living Init containers are covered
```
---
apiVersion: v1
kind: Pod
metadata:
  name: mytarget
  labels:
    app: test
spec:
  initContainers:
    - name: test-init
      image: ubuntu:latest
      command:
        - bash
        - -c
        - |
          #!/bin/bash
          echo "$(date): started the INIT process"
          sleep 10
          echo "$(date): sleeping 10 INIT seconds"
  containers:
    - name: test
      image: ubuntu:latest
      command:
        - bash
        - -c
        - |
          #!/bin/bash
          echo "$(date): started the process"

          while :
          do
                 echo "$(date): sleeping 5 seconds"
                 sleep 5
          done
```

### Ensure logs are captured even if Pod is marked for deletion
```
---
apiVersion: v1
kind: Pod
metadata:
  name: mytarget2
  labels:
    app: test
spec:
  containers:
    - name: test
      image: ubuntu:latest
      command:
        - bash
        - -c
        - |
          #!/bin/bash
          trap "{ echo '$(date): SIGTERM triggered'; sleep 3;echo '$(date): Bye bye'; sleep 5; echo '$(date): bye'; sleep 5; exit 1; }" SIGINT SIGTERM
          echo "$(date): started the process"

          while :
          do
                 echo "$(date): sleeping 5 seconds"
                 sleep 5
          done
```

## Related issues
- Closes https://github.com/elastic/beats/issues/24495, https://github.com/elastic/beats/issues/27019

## Notes for the reviewer
Most of the functionality is ported from `libbeat/autodiscover/providers/kubernetes` and `libbeat/common/kubernetes/metadata`.

### CODE REVIEW

**Comment by ChrsMark** [comment]
I'm opening this for an early review. I will need to add/tune tests as well as do extensive manual testing, however an early review would be more than welcome so as to spot any possible foundational issues early on.

**Comment by ChrsMark** [comment]
Heads-up on this:

Unit tests were tuned accordingly and I'm running several manual testing scenarios to verify nothing is broken. I'm planning to work on adding e2e tests right after this one is merged (issue: https://github.com/elastic/e2e-testing/issues/1090)

I have particularly tried to "port" the logic from the old autodiscovery feature to this one and try to cover cases like short-living jobs etc. 

Review parts:
- @jsoriano it would be super helpful if you could review the discovery lifecycle start/stop events , proper termination etc since you had worked on this recently :)
- @blakerouse could you review this patch from the Agent's framework perspective? 
- @exekias one more review since you initially shipped the k8s provider :) ?
- @MichaelKatsoulis the provider will not add `kubernetes.container.image` but you can verify the ECS parts, and feel free to review the patch in general.


Maybe in the same release we can tune the metadata part (in follow-ups) in order to be aligned with the outcomes of https://github.com/elastic/beats/issues/13911 and more specifically https://github.com/elastic/beats/issues/16483 and https://github.com/elastic/beats/pull/14875

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go:36]: Can this just be the default? Do we need this configurable now?

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go:39]: If undocumented and deprecated, why add it? Being this is all new to Elastic Agent, it could be the time to remove it.

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go:33]: Where is `IncludeLabels` and `ExcludeLabels` used? I was not able to find them in the diff.

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/kubernetes.go
Comment (jsoriano) [inline x-pack/elastic-agent/pkg/composable/providers/kubernetes/kubernetes.go:13]: Nit. 
```suggestion
```

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/node.go
Comment (jsoriano) [inline x-pack/elastic-agent/pkg/composable/providers/kubernetes/node.go:11]: Nit. Move beats imports to the last group of imports.

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/kubernetes.go
Comment (jsoriano) [inline x-pack/elastic-agent/pkg/composable/providers/kubernetes/kubernetes.go:140]: It looks a bit weird to me to need to pass another config, raw, to this method. Some thoughts about this, no need to do anything now.

The `Config` type of `config` is quite similar to the `metadata.Config` type used by the underlying metadata generator. I guess that the new Config object was introduced here to decouple the composable provider config, but at this point they are becoming more similar, I wonder if they could became the same thing.

Talking about decoupling, we are implementing another set of watchers here, would it make sense to also implement a new set of metadata generators here? I am totally +1 with reusing them, but I wonder if this would be a chance to clean something.

Other thing that could be explored could be to refactor `metadata` package, so the constructors there receive a `metadata.Config` instead of a raw `common.Config`. This would allow to more cleanly reuse them, for example here it would allow to generate or unpack this metadata config here, and pass it to the watchers, instead of the raw config.

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/pod.go
Comment (jsoriano) [inline x-pack/elastic-agent/pkg/composable/providers/kubernetes/pod.go:485]: There are two places now where we collect all the containers in the pod and their statuses (the other [here](https://github.com/elastic/beats/blob/848c275cafb1b1ac8a26f58bb6eb0f266cadc3d3/x-pack/elastic-agent/pkg/composable/providers/kubernetes/pod.go#L139)), this can be error-prone in future refactors. I wonder if this could be unified.

**Review by jsoriano — COMMENTED** [review]
> @jsoriano it would be super helpful if you could review the discovery lifecycle start/stop events , proper termination etc since you had worked on this recently :)

This part looks good (waiting for e2e tests for confirmation :slightly_smiling_face:), added only some thoughts, nothing really blocking.

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go
Comment (ChrsMark) [reply x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go:36]: I'm not sure about the history of these settings. Maybe some users had requests for these? @jsoriano @exekias do you think we could remove those settings and have dedoting as an always-on feature?

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go
Comment (ChrsMark) [reply x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go:39]: You are right this one is not exposed so it should be removed.

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go
Comment (ChrsMark) [reply x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go:33]: These are passed and used deeper in meta Generators, which is already existent codebase.

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/kubernetes.go
Comment (ChrsMark) [reply x-pack/elastic-agent/pkg/composable/providers/kubernetes/kubernetes.go:140]: I cahnge the code a little bit in order to avoid moving around the configs (both basic and raw). 


As far as the decoupling is concerned I think that it's not such a good idea since meta Generators are already used by several other parts like `add_k8s_metadata`, k8s module and k8s autodiscover (beats) so it stands as a single source of truth for metadata. Since the porcessor and the module are coming along to Agent's era it would be nice to preserve this common ground for them and have a single metadata generation library. Just my personal point of view but so far it has been quite useful to handle meta from a single place, happy to discuss it more though if there is a better approach. 


Regarding the cleanups/reorganisation of generators I'm totally +1 but I would wait to see how the projects will evolve in the near future and work on cleanups while we move (I still have hopes that hints are coming soon :)).

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/pod.go
Comment (ChrsMark) [reply x-pack/elastic-agent/pkg/composable/providers/kubernetes/pod.go:485]: Fair point. I will tune it! :)

**Comment by ChrsMark** [comment]
@jsoriano @blakerouse I tested this extensively (manually) and I think is good to go. Would you mind giving it another review?

Next steps will be to add 2e2 tests (in this iteration/release) and verify that we are aligned with #13911 (cc @MichaelKatsoulis ).

**Review by blakerouse — APPROVED** [review]
Overall this looks good to me.

I would prefer to see the `dedot` feature just set to default and the options to turn if off removed. But I will leave that up to you to determine.

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go
Comment (ChrsMark) [reply x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go:39]: Actually this touches code that is used by beats too. I will remove it in follow up PR target 8.0

**Review by jsoriano — APPROVED** [review]

File: x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go
Comment (ChrsMark) [reply x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go:38]: PR: https://github.com/elastic/beats/pull/28006

**Pull request merged**
